### PR TITLE
add tests and permissions for event app

### PR DESCRIPTION
**Pull request by arminpatel**
## Overview

1. This PR fixes or fixes part of #[fill_in_number_here].
2. This PR does the following: [Explain here what your PR does and why]

## Essential Checklist

- [ ] The PR title starts with "Fix #bugnum: ", followed by a short, clear summary of the changes. (If this PR fixes part of an issue, prefix the title with "Fix part of #bugnum: ...".)
- [ ] "Allow edits from maintainers" is checked. (See [here](https://help.github.com/en/github/collaborating-with-issues-and-pull-requests/allowing-changes-to-a-pull-request-branch-created-from-a-fork) for instructions on how to enable it.)
- [ ] The PR is made from a branch that's **not** called "main/master".

## PR Pointers

- If you need a review or an answer to a question, and don't have permissions to assign people, **leave a comment** like the following: "{{Question/comment}} @{{reviewer_username}} PTAL".
- Never force push. If you do, your PR will be closed.

### CODE REVIEW

File: backend/events/tests.py [hunk 1]
```diff
@@ -1,3 +1,237 @@
...
+    event_data_copy = event_data.copy()
+    event_data_copy['organizer'] = test_user_2
+    event = Event.objects.create(**event_data_copy)
+
+    # Act
+    response = authenticated_client.delete(f'/api/events/{event.id}/')
+
+    # Assert
+    assert response.status_code == status.HTTP_403_FORBIDDEN
+    assert Event.objects.count() == 1
```
Comment (Priyansh61) [inline backend/events/tests.py:226]: better way ig

`assert Event.objects.filter(id=event.id).exists()
`

File: backend/events/tests.py [hunk 2]
```diff
@@ -1,3 +1,237 @@
...
+
+@pytest.mark.django_db
+def testDeleteEventView_invalidEventId_shouldThrowNotFound(authenticated_client, test_user, test_event):
+    # Arrange
+    # Act
+    response = authenticated_client.delete(f'/api/events/{test_event.id + 1}/')
+
+    # Assert
+    assert response.status_code == status.HTTP_404_NOT_FOUND
+    assert Event.objects.count() == 1
```
Comment (Priyansh61) [inline backend/events/tests.py:237]: same as previous

**Review by Priyansh61 — COMMENTED** [review]
And please check the function naming try to make them more assertive if possible.

File: backend/events/tests.py [hunk 3]
```diff
@@ -1,3 +1,237 @@
...
+    event_data_copy = event_data.copy()
+    event_data_copy['organizer'] = test_user.id
+
+    # Act
+    response = authenticated_client.post('/api/events/', event_data_copy, format='json')
+
+    print(response.data)
+    # Assert
+    assert response.status_code == status.HTTP_201_CREATED
+    assert Event.objects.count() == 1
```
[the remark this answers is not in the document]
Comment (arminpatel) [reply backend/events/tests.py:61]: won't matter as `@pytest.mark.django_db` ensures that a clean database is used for each test

File: backend/events/tests.py [hunk 4]
```diff
@@ -1,3 +1,237 @@
...
+@pytest.mark.django_db
+def testCreateEventView_validEventDetails_authenticatedUser_shouldCreateEvent(test_user, authenticated_client):
+    # Arrange
+    event_data_copy = event_data.copy()
+    event_data_copy['organizer'] = test_user.id
+
+    # Act
+    response = authenticated_client.post('/api/events/', event_data_copy, format='json')
+
+    print(response.data)
```
[the remark this answers is not in the document]
Comment (arminpatel) [reply backend/events/tests.py:58]: left by mistake, will remove it

File: backend/events/tests.py [hunk 5]
```diff
@@ -1,3 +1,237 @@
...
+    # Arrange
+    unauthenticated_client = APIClient()
+    event_data_copy = event_data.copy()
+    event_data_copy['organizer'] = test_user.id
+
+    # Act
+    response = unauthenticated_client.post('/api/events/', event_data_copy, format='json')
+
+    # Assert
+    assert response.status_code == status.HTTP_403_FORBIDDEN
```
[the remark this answers is not in the document]
Comment (arminpatel) [reply backend/events/tests.py:103]: [reference](https://www.django-rest-framework.org/api-guide/authentication/#unauthorized-and-forbidden-responses)

File: backend/events/tests.py [hunk 6]
```diff
@@ -1,3 +1,237 @@
...
+def testRetrieveEventView_validEventId_shouldRetrieveEvent(test_user, test_event):
+    # Arrange
+    unauthenticated_client = APIClient()
+
+    # Act
+    response = unauthenticated_client.get(f'/api/events/{test_event.id}/')
+
+    # Assert
+    assert response.status_code == status.HTTP_200_OK
+    assert response.data['title'] == event_data['title']
```
[the remark this answers is not in the document]
Comment (arminpatel) [reply backend/events/tests.py:117]: sure

File: backend/events/tests.py [hunk 7]
```diff
@@ -1,3 +1,237 @@
...
+
+
+@pytest.mark.django_db
+def testRetrieveEventView_invalidEventId_shouldThrowNotFound(test_user, test_event):
```
[the remark this answers is not in the document]
Comment (arminpatel) [reply backend/events/tests.py:121]: yeah not needed, will remove it

File: backend/events/tests.py [hunk 8]
```diff
@@ -1,3 +1,237 @@
...
+    # Arrange
+    unauthenticated_client = APIClient()
+
+    # Act
+    response = unauthenticated_client.get('/api/events/')
+
+    # Assert
+    assert response.status_code == status.HTTP_200_OK
+    assert len(response.data) == 1
+    assert response.data[0]['title'] == event_data['title']
```
[the remark this answers is not in the document]
Comment (arminpatel) [reply backend/events/tests.py:143]: sure

File: backend/events/tests.py [hunk 9]
```diff
@@ -1,3 +1,237 @@
...
+    # Arrange
+    event_data_copy = event_data.copy()
+    event_data_copy['title'] = 'updated title'
+    event_data_copy['organizer'] = test_user.id
+    event_data_copy['date'] = '2021-01-02'
+
+    # Act
+    response = authenticated_client.put(f'/api/events/{test_event.id}/', event_data_copy, format='json')
+
+    # Assert
```
[the remark this answers is not in the document]
Comment (arminpatel) [reply backend/events/tests.py:157]: okay, let's assert that too

File: backend/events/tests.py [hunk 10]
```diff
@@ -1,3 +1,237 @@
...
+    event_data_copy = event_data.copy() 
+    event_data_copy['title'] = 'updated title'
+    event_data_copy['organizer'] = test_user.id
+
+    # Act
+    response = authenticated_client.put(f'/api/events/{test_event.id + 1}/', event_data_copy, format='json')
+
+    # Assert
+    assert response.status_code == status.HTTP_404_NOT_FOUND
+    assert Event.objects.get().title == event_data['title']
```
[the remark this answers is not in the document]
Comment (arminpatel) [reply backend/events/tests.py:196]: Is there any particular reason to favor this?

File: backend/events/tests.py (hunk 1 above)
Comment (arminpatel) [reply backend/events/tests.py:226]: why though?

**Comment by arminpatel** [comment]
more assertive names sound good, I would just add some camel casing. It makes them much more readable imo

File: backend/events/tests.py (hunk 10 above)
[the remark this answers is not in the document]
Comment (Priyansh61) [reply backend/events/tests.py:196]: I just read it somewhere do take a look that it is agreat way to check the db after filling any put or patch.

Just check it once again on your end

File: backend/events/tests.py (hunk 1 above)
Comment (Priyansh61) [reply backend/events/tests.py:226]: Just a better way to do it I would say

File: backend/events/tests.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (Priyansh61) [reply backend/events/tests.py:121]: It is also there at some place so please check them too

File: backend/events/tests.py [hunk 11]
```diff
@@ -1,3 +1,237 @@
...
+    assert response.data['title'] == event_data['title']
+
+
+@pytest.mark.django_db
+def testRetrieveEventView_invalidEventId_shouldThrowNotFound(test_user, test_event):
+    # Arrange
+    unauthenticated_client = APIClient()
+
+    # Act
+    response = unauthenticated_client.get(f'/api/events/{test_event.id + 1}/')
```
[the remark this answers is not in the document]
Comment (arminpatel) [reply backend/events/tests.py:140]: I don't think it's needed as there's only 1 event in the db

File: backend/events/tests.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (arminpatel) [reply backend/events/tests.py:104]: in drf's implementation, when object level permissions are evaluated and denied a 403 is thrown

**Comment by Priyansh61** [comment]
@arminpatel PTAL!

**Review by Priyansh61 — APPROVED** [review]

**Pull request merged**